File: DSAlgorithms/Dynamic_Programming/c++/gridTraveler.cpp
#include<iostream>
#include<unordered_map>

int gridTraveler(int, int);
int gridTravelerMemoized(int, int, std::unordered_map<std::string, int>);
int main()
{
    std::cout<<gridTraveler(2,3)<<'\n';
    std::unordered_map<std::string, int> cache;
    std::cout<<gridTravelerMemoized(2,3, cache)<<'\n';
    return 0;
}
int gridTraveler(int m , int n)
{
    if(m == 1 && n == 1)
        return 1;
    if(m == 0 || n == 0)
        return 0;
    return gridTraveler(m - 1, n) + gridTraveler(m, n - 1);
}


int gridTravelerMemoized(int m , int n, std::unordered_map<std::string, int> cache)
{
    std::string key = std::to_string(m) + ',' + std::to_string(n);
    if(cache[key])
        return cache[key];
    if(m == 1 && n == 1)
        return 1;
    if(m == 0 || n == 0)
        return 0;
    cache[key] = gridTravelerMemoized(m - 1, n, cache) + gridTravelerMemoized(m, n - 1, cache);
    return cache[key];
}

File: dsa_algo/topics/Dynamic_Programming/ts/src/gridTraveler.ts
const gridTraveler = (m: number, n: number): number => {
    if(m === 1 && n == 1)
        return 1;
    if(m === 0 || n === 0)
        return 0;
    return gridTraveler(m -1, n) + gridTraveler(m, n - 1);
}

const gridTravelerMemoized = (m: number, n: number): number => {
    const cache = new Map<string, number>();
    const key = m + ',' + n;
    const checkKey = cache.get(key);
    if(checkKey)
        return checkKey;
    if(m === 1  && n === 1)
        return 1;
    if(m === 0 || n === 0)
        return 0;
    cache.set(key, gridTravelerMemoized(m - 1, n) + gridTravelerMemoized(m, n - 1));
    let currentVal: number = 0;
    const getCurrentVal = cache.get(key);
    if(getCurrentVal)
        currentVal = getCurrentVal;
    return currentVal;
}

console.log(gridTraveler(2,3));
console.log(gridTravelerMemoized(2,3));
